homeassistant/components/homekit/type_sensors.py

```diff
@@ -128,21 +153,244 @@ def __init__(self, *args):
         """Initialize a AirQualitySensor accessory object."""
         super().__init__(*args, category=CATEGORY_SENSOR)
 
-        serv_air_quality = self.add_preload_service(
-            SERV_AIR_QUALITY_SENSOR, [CHAR_AIR_PARTICULATE_DENSITY]
+        """determine the charciteristics of the sensor based on the curent attributes"""
+        chars = []
+        current_state = self.hass.states.get(self.entity_id)
+        nitrogen_dioxide_density = convert_to_float(
+            current_state.attributes.get(ATTR_NITROGEN_DIOXIDE_DENSITY)
+        )
+        voc_density = convert_to_float(current_state.attributes.get(ATTR_VOC_DENSITY))
+        pm_2_5_density = convert_to_float(
+            current_state.attributes.get(ATTR_PM_2_5_DENSITY)
         )
-        self.char_quality = serv_air_quality.configure_char(CHAR_AIR_QUALITY, value=0)
-        self.char_density = serv_air_quality.configure_char(
-            CHAR_AIR_PARTICULATE_DENSITY, value=0
+        pm_10_density = convert_to_float(
+            current_state.attributes.get(ATTR_PM_10_DENSITY)
+        )
+        pm_density = convert_to_float(current_state.attributes.get(ATTR_PM_DENSITY))
+        pm_size = convert_to_float(current_state.attributes.get(ATTR_PM_SIZE))
+
+        if nitrogen_dioxide_density is not None:
+            chars.append(CHAR_NITROGEN_DIOXIDE_DENSITY)
+
+        if voc_density is not None:
+            chars.append(CHAR_VOC_DENSITY)
+
+        """
+        NOTE: this would be the easy way but it seems the driver is missing the PM2_5_Density key
+              because trying to do this causes the following when tyring to call self.add_preload_service:
+
+                File "/usr/local/lib/python3.7/site-packages/pyhap/loader.py", line 42, in get_char
+                  char_dict = self.char_types[name].copy()
+              KeyError: 'PM2_5Density'
+
+        if pm_2_5_density is not None:
+            chars.append(CHAR_PM_2_5_DENSITY)
+        """
+
+        if pm_10_density is not None:
+            chars.append(CHAR_PM_10_DENSITY)
+
+        if (
+            (pm_density is not None)
+            and (pm_2_5_density is None)
+            and (pm_10_density is None)
+        ):
+            chars.append(CHAR_AIR_PARTICULATE_DENSITY)
+        elif (pm_density is not None) and (
+            (pm_2_5_density is not None) or (pm_10_density is not None)
+        ):
+            _LOGGER.warn(
+                "%s: Has pm_density and (pm_2_5_density and/or pm_10_density) these are mutually exclusive charicteristics so ignoring pm_density",
```

Comment:
AIUI apple added AirPariculateSize and AirPariculateDensity in iOS9, and with those you couldn't have both PM2.5 and PM10 values at the same time. With iOS10.3 They added seperate PM2.5 and PM10 fields. In both public versions of the HAP spec the Air Quality Sensor 'service' has listed PM2.5 and PM10 as optional characteristics and not listed the AirParitculate* characteristics at all.

So it kind of feels like the characteristics we currently use might be out of favour for new devices?

I wonder then if we should just drop `CHAR_AIR_PARTICULATE_DENSITY` and `CHAR_AIR_PARTICULATE_SIZE` rather than try to support both characteristic types? I suggest it because it would simplify this code a fair bit AND we won't be adding warnings to our users setups.

Can you think of any reasons why not do do that?

Reply:
long story short, i thought the PM2.5/PM10 characteristic was somehow mutually exclusive with the PM Density and PM Size characteristics, so the code got complicated because of that. But turns out they aren't actually exclusive and you can set them all, so now all the confusing complicated code is gone. I would think just leave them all in as options now that it's not complicated to do so. Mostly so we can maintain backwards compatibility with the old code that assumed the AirQuality sensor was only for PM2.5 and the functionally that provided.